File: bms/go/triangolatte/combinePolygons/combinePolygons.go
package test

import (
	"container/list"
	"errors"
)

var ()

const ()

func combinePolygons(outer, inner []Point) ([]Point, error) {
	xMax := 0.0
	mIndex := 0
	for i := 0; i < len(inner); i++ {
		if inner[i].X > xMax {
			xMax = inner[i].X
			mIndex = i
		}
	}

	m := inner[mIndex]

	var pIndex int
	visibleIndex := -1

	k, k1, k2, err := findK(m, outer)

	if err != nil {
		return nil, err
	}

	// If `K` is vertex of the outer polygon, `M` and `K` are mutually visible.
	for i := 0; i < len(outer); i++ {
		if outer[i] == k {
			visibleIndex = i
		}
	}

	// Otherwise, `K` is an interior point of the edge `[V_k_1, V_k_2]`. Find `P`
	// which is endpoint with greater x-value.
	if outer[k1].X > outer[k2].X {
		pIndex = k1
	} else {
		pIndex = k2
	}

	// Check with all vertices of the outer polygon to be outside of the
	// triangle `[M, K, P]`. If it is true, `M` and `P` are mutually visible.
	allOutside := areAllOutside(m, k, pIndex, outer)

	if visibleIndex < 0 && allOutside {
		visibleIndex = pIndex
	}

	// Otherwise at least one reflex vertex lies in `[M, K, P]`. Search for the
	// array of reflex vertices `R` that minimizes the angle between `(1, 0)` and
	// line segment `[M, R]`. If there is exactly one vertex in `R` then they are
	// mutually visible. If there are multiple such vertices, pick the one closest
	// to `M`.
	if visibleIndex < 0 {
		visibleIndex = findClosest(m, k, pIndex, outer)
	}

	if visibleIndex < 0 {
		return nil, errors.New("could not find visible vertex")
	}

	result := make([]Point, 0, len(outer)+len(inner)+2)
	result = append(result, outer[:visibleIndex+1]...)
	for i := 0; i < len(inner); i++ {
		result = append(result, inner[cyclic(mIndex+i, len(inner))])
	}
	result = append(result, inner[mIndex], outer[visibleIndex])
	result = append(result, outer[visibleIndex+1:]...)

	return result, nil
}

func findK(m Point, outer []Point) (k Point, k1, k2 int, err error) {
	for i, j := len(outer)-1, 0; j < len(outer); i, j = j, j+1 {
		// Skip edges that does not have their first point below `M` and the second
		// one above.
		if outer[i].Y > m.Y || outer[j].Y < m.Y {
			continue
		}

		// Calculate simplified intersection of ray (1, 0) and [V_i, V_j] segment.
		v1 := m.Sub(outer[i])
		v2 := outer[j].Sub(outer[i])

		t1 := v2.Cross(v1) / v2.Y
		t2 := v1.Y / v2.Y

		if t1 >= 0.0 && t2 >= 0.0 && t2 <= 1.0 {
			// If there is no current `k` candidate or this one is closer.
			if t1-m.X < k.X {
				k = Point{X: t1 + m.X, Y: m.Y}
				k1, k2 = i, j
				return
			}
		} else {
			err = errors.New("cannot calculate intersection, problematic data")
			return
		}
	}
	return
}

func (p Point) Sub(r Point) Point {
	return Point{p.X - r.X, p.Y - r.Y}
}

func (p Point) Cross(r Point) float64 {
	return p.X*r.Y - p.Y*r.X
}

func areAllOutside(m, k Point, pIndex int, outer []Point) bool {
	allOutside := true
	for i := range outer {
		// We have to skip M, K and P vertices. Since M is from the inner
		// polygon and K was proved to not match any vertex, the only one to
		// check is pIndex
		if i == pIndex {
			continue
		}

		if isInsideTriangle(m, k, outer[pIndex], outer[i]) {
			allOutside = false
		}
	}
	return allOutside
}

func isInsideTriangle(a, b, c, p Point) bool {
	return (c.X-p.X)*(a.Y-p.Y)-(a.X-p.X)*(c.Y-p.Y) >= 0 &&
		(a.X-p.X)*(b.Y-p.Y)-(b.X-p.X)*(a.Y-p.Y) >= 0 &&
		(b.X-p.X)*(c.Y-p.Y)-(c.X-p.X)*(b.Y-p.Y) >= 0
}

func findClosest(m, k Point, pIndex int, outer []Point) int {
	reflex := list.New()
	n := len(outer)
	for i := 0; i < n; i++ {
		notInside := !isInsideTriangle(m, k, outer[pIndex], outer[i])
		prev, next := cyclic(i-1, n), cyclic(i+1, n)
		notReflex := !isReflex(outer[prev], outer[i], outer[next])
		if notInside || notReflex {
			continue
		}
		reflex.PushBack(i)
	}
	var closest int
	var maxDist float64

	for r := reflex.Front(); r != nil; r = r.Next() {
		i := r.Value.(int)
		dist := outer[i].Distance2(outer[closest])
		if dist > maxDist {
			closest = i
			maxDist = dist
		}
	}
	return closest
}

func cyclic(i, n int) int {
	return (i%n + n) % n
}

func isReflex(a, b, c Point) bool {
	return (b.X-a.X)*(c.Y-b.Y)-(c.X-b.X)*(b.Y-a.Y) < 0
}

func (p Point) Distance2(r Point) float64 {
	return (p.X-r.X)*(p.X-r.X) + (p.Y-r.Y)*(p.Y-r.Y)
}

type Point struct {
	X, Y float64
}

File: translations_only/fail/go/triangolatte/combinePolygons/claude3/CAPR/combinePolygons.rs


use std::collections::LinkedList;

#[derive(Debug, Copy, Clone, PartialEq)]
struct Point {
    x: f64,
    y: f64,
}

fn sub(p: Point, r: Point) -> Point {
    Point {
        x: p.x - r.x,
        y: p.y - r.y,
    }
}

fn cross(p: Point, r: Point) -> f64 {
    p.x * r.y - p.y * r.x
}

fn combine_polygons(outer: &[Point], inner: &[Point]) -> Result<Vec<Point>, &'static str> {
    let mut x_max = 0.0;
    let mut m_index = 0;
    for i in 0..inner.len() {
        if inner[i].x > x_max {
            x_max = inner[i].x;
            m_index = i;
        }
    }

    let m = inner[m_index];

    let (mut k, mut k1, mut k2) = (Point { x: 0.0, y: 0.0 }, 0, 0);
    if let Err(err) = find_k(m, outer, &mut k, &mut k1, &mut k2) {
        return Err(err);
    }

    let mut visible_index = -1;

    for i in 0..outer.len() {
        if outer[i] == k {
            visible_index = i as isize;
        }
    }

    let p_index = if outer[k1].x > outer[k2].x {
        k1
    } else {
        k2
    };

    let all_outside = are_all_outside(m, k, p_index, outer);

    if visible_index < 0 && all_outside {
        visible_index = p_index as isize;
    }

    if visible_index < 0 {
        visible_index = find_closest(m, k, p_index, outer) as isize;
    }

    if visible_index < 0 {
        return Err("could not find visible vertex");
    }

    let mut result = Vec::with_capacity(outer.len() + inner.len() + 2);
    result.extend_from_slice(&outer[..=(visible_index as usize)]);
    for i in 0..inner.len() {
        result.push(inner[cyclic(m_index as isize + i as isize, inner.len() as isize) as usize]);
    }
    result.push(inner[m_index]);
    result.push(outer[visible_index as usize]);
    result.extend_from_slice(&outer[(visible_index as usize + 1)..]);

    Ok(result)
}

fn find_k(
    m: Point,
    outer: &[Point],
    k: &mut Point,
    k1: &mut usize,
    k2: &mut usize,
) -> Result<(), &'static str> {
    for (i, j) in (0..outer.len()).zip(1..outer.len()).chain(Some((outer.len() - 1, 0))) {
        if outer[i].y > m.y || outer[j].y < m.y {
            continue;
        }

        let v1 = sub(m, outer[i]);
        let v2 = sub(outer[j], outer[i]);

        let t1 = cross(v2, v1) / v2.y;
        let t2 = v1.y / v2.y;

        if t1 >= 0.0 && t2 >= 0.0 && t2 <= 1.0 {
            if t1 - m.x < k.x {
                k.x = t1 + m.x;
                k.y = m.y;
                *k1 = i;
                *k2 = j;
                return Ok(());
            }
        } else {
            return Err("cannot calculate intersection, problematic data");
        }
    }
    Ok(())
}

fn are_all_outside(m: Point, k: Point, p_index: usize, outer: &[Point]) -> bool {
    let mut all_outside = true;
    for i in 0..outer.len() {
        if i == p_index {
            continue;
        }

        if is_inside_triangle(m, k, outer[p_index], outer[i]) {
            all_outside = false;
        }
    }
    all_outside
}

fn is_inside_triangle(a: Point, b: Point, c: Point, p: Point) -> bool {
    (c.x - p.x) * (a.y - p.y) - (a.x - p.x) * (c.y - p.y) >= 0.0
        && (a.x - p.x) * (b.y - p.y) - (b.x - p.x) * (a.y - p.y) >= 0.0
        && (b.x - p.x) * (c.y - p.y) - (c.x - p.x) * (b.y - p.y) >= 0.0
}

fn find_closest(m: Point, k: Point, p_index: usize, outer: &[Point]) -> usize {
    let mut reflex = LinkedList::new();
    let n = outer.len();
    for i in 0..n {
        let not_inside = !is_inside_triangle(m, k, outer[p_index], outer[i]);
        let prev = cyclic(i as isize - 1, n as isize) as usize;
        let next = cyclic(i as isize + 1, n as isize) as usize;
        let not_reflex = !is_reflex(outer[prev], outer[i], outer[next]);
        if not_inside || not_reflex {
            continue;
        }
        reflex.push_back(i);
    }
    let mut closest = 0;
    let mut max_dist = 0.0;

    for r in reflex.iter() {
        let i = *r;
        let dist = distance2(outer[i], outer[closest]);
        if dist > max_dist {
            closest = i;
            max_dist = dist;
        }
    }
    closest
}

fn cyclic(i: isize, n: isize) -> isize {
    ((i % n + n) % n)
}

fn is_reflex(a: Point, b: Point, c: Point) -> bool {
    (b.x - a.x) * (c.y - b.y) - (c.x - b.x) * (b.y - a.y) < 0.0
}

fn distance2(p: Point, r: Point) -> f64 {
    (p.x - r.x) * (p.x - r.x) + (p.y - r.y) * (p.y - r.y)
}

